### ISSUE TITLE
[ADR] Segregate execution vs research into two logical paths → beacon_core + beacon_research (two-tier deploy at go-live)

### ISSUE THREAD

**Issue by magedzamzam** [magedzamzam/beacon-trader#60]
## Status

**RECORDED — deferred.** This is an architecture decision we want on the record now and execute later. **Do not implement yet.** Trigger to act: **funding a live account**, or research query-load visibly contending with the `monitor` (whichever comes first). We are on a DEMO eval account today, so the cost of a premature split outweighs the benefit — see [eval context].

## Context

Everything ships as one package (`beacon_core`) baked into **every** Python image, so a change to a research module (e.g. `analysis/bayes.py`) forces a rebuild of the trading services (`executor`, `monitor`). That couples two workloads with opposite requirements:

- **Execution path** — sizes orders, moves stops, cancels/expires entries. Must be **lean, always-on, low-latency, minimal deps, broker-write**. Rarely changed.
- **Research path** — Bayesian edge, reconciler, performance, PM-bot. **Heavy, batchy, read-only, changes constantly.** Downtime is harmless (already designed to degrade gracefully).

We already run research on the **daily `pg_dump`, not the live DB** — so the "research reads a copy, never touches prod" pattern exists in practice; this ADR makes it the intended architecture and prepares the package seam.

## Decision

1. **Define two logical paths now**, one physical deployment. Keep research strictly read-only against a copy (dump/replica); research influences production **only through reviewed config/PRs**.
2. **Carve `beacon_core` into three packages** so the boundary is enforced in code (not just convention):
   - `beacon_common` — shared kernel (db/models, config, settings, crypto, security, logging, bus, health, timeutil, tasks).
   - `beacon_core` — the execution/money path. **May depend on `beacon_common` only.**
   - `beacon_research` — analysis. **May depend on `beacon_common`; MUST NOT be imported by `beacon_core`.**
3. **Physically split at the trigger** into a lean Production tier (broker-write) and a heavy Research tier (read replica, no broker access).

### Hard rule (the whole point)
> `beacon_core` (execution) must **never** import `beacon_research`. Data flows **one way**: prod DB → replica → research. Research → prod happens **only** as reviewed configuration/PRs.

## Target topology

```mermaid
flowchart TB
  subgraph PROD["PRODUCTION TIER — EXECUTION (lean · always-on · broker-write)"]
    direction LR
    TG["telegram<br/>ingest"]
    COL["collector<br/>market / claims feed"]
    APIC["api<br/>config + ingest webhooks"]
    RDS["redis<br/>bus"]
    EXE["executor<br/>parse · plan · size · guard · place"]
    MON["monitor<br/>reconcile · SL rules · TTL/GTD"]
    FE["frontend<br/>config + ops"]
    BRK{{"Broker — Capital.com"}}
    PGDB[("Postgres<br/>ledger — source of truth")]
    TG --> APIC
    COL --> APIC
    APIC -->|validated signal| RDS
    RDS --> EXE
    EXE --> BRK
    BRK --> MON
    EXE --> PGDB
    MON --> PGDB
    FE --- APIC
  end

  subgraph RES["RESEARCH TIER — ANALYSIS (heavy · on-demand · read-only)"]
    direction LR
    REP[("read replica<br/>/ daily pg_dump")]
    RAPI["research-api<br/>performance · analytics · reconciliation"]
    BZ["bayes · estimators<br/>edge / correlation"]
    RC["reconciler<br/>channel-claim vs fill"]
    TAR["ta analysis<br/>feature study"]
    PM["PM-bot<br/>daily roadmaps + issues"]
    REP --> RAPI
    REP --> BZ
    REP --> RC
    REP --> TAR
    REP --> PM
  end

  PGDB -. "one-way: streaming replica / nightly dump" .-> REP
  PM -. "influence prod ONLY via reviewed config + PRs" .-> FE
```

## Package dependency layering

```mermaid
flowchart TB
  subgraph L3["beacon_research — ANALYSIS"]
    R1["analysis: bayes · estimators · reconcile · report · sidecar · claims-compute<br/>ta: features · indicators (study) · alpha · research API routers"]
  end
  subgraph L2["beacon_core — EXECUTION (money path)"]
    E1["brokers · parsing · ingest · execution/planner+guard<br/>risk/sizing · strategy/rules · trading_hours · notifications · ai (thin exec gate, optional)"]
    E2["thin producers: ta/capture · analysis/claims-capture<br/>(run inside exec services, write to DB only)"]
  end
  subgraph L1["beacon_common — shared kernel"]
    C1["db/models · config · settings_store · crypto · security<br/>logging · bus · health · timeutil · tasks"]
  end
  L2 --> L1
  L3 --> L1
  L3 -. "reads ledger via replica — NEVER imported by beacon_core" .-> L1
```

## Module → target package

| Current module | Target package | Notes |
|---|---|---|
| `db/`, `config.py`, `settings_store.py`, `crypto.py`, `security.py`, `logging.py`, `bus.py`, `health.py`, `timeutil.py`, `tasks.py` | **beacon_common** | shared kernel |
| `brokers/`, `parsing/`, `ingest/`, `execution/`, `risk/`, `strategy/`, `trading_hours/`, `notifications/` | **beacon_core** | the money path |
| `ai/` | **beacon_core** (thin gate) | optional/advisory; keep the inline gate thin, defer any heavy prompt work |
| `ta/capture`, `analysis/claims` (capture side) | **beacon_core** (producers) | run in exec services, write `SignalFeature` / claims to DB only |
| `analysis/` (bayes, estimators, reconcile, report, sidecar), `ta/` (features, indicators, registry — study), `alpha/` | **beacon_research** | heavy compute |
| API routers: `analysis`, `analytics`, `entry_filters`, `performance`, `reconciliation`, `ta` | **beacon_research** (research-api) | read/dashboard surfaces |

## Coupling to unwind (the concrete refactor when we execute)

Today the execution path imports research directly — these are the seams to convert into thin DB-producers:
- `services/executor/main.py` → imports `beacon_core.ta` (feature capture). Keep a **thin `ta/capture`** producer in core; move feature *analysis* to research.
- `services/monitor/main.py` → imports `beacon_core.analysis` (reconcile). Monitor should only **write** reconciliation inputs; the compute moves to research.
- `services/telegram/main.py` → imports `beacon_core.analysis` (claims). Same: **capture** the claim in core, **analyse** in research.

## Migration phases

1. **Now (this ADR):** record the decision + topology. No code change. Keep enforcing "research reads the dump, never writes the trading path."
2. **Package carve (do first, low-risk, single deploy):** split into `beacon_common` / `beacon_core` / `beacon_research`; fix the 3 producer imports so `beacon_core` no longer imports research. **Payoff even without a second env:** research changes stop triggering execution-image rebuilds.
3. **Physical split (at trigger):** stand up the Research tier against a read replica; strip research images/creds from the Production tier; Production keeps broker-write only.

## Consequences

**Pros:** execution blast-radius isolated from heavy analytics; independent deploy cadence; least-privilege on the money path (research has no broker access); research can scale/iterate freely.
**Cons/cost:** a replica or dump pipeline to maintain; two deploys at the final stage; a one-time package refactor. Mitigated by doing phase 2 (package carve) first — most of the benefit, one deployment.

## Non-goals

- Not changing trade logic. Not splitting the DB (single source of truth stays in Production; research reads a copy). Not doing this before go-live.

## Related

- #34 (Redis pub/sub reliability), #35 (BaseInboundChannel), #36 (shared-infra helpers) — same architecture-hygiene thread; the package carve here subsumes part of #36.
- #40 (order GTD/TTL) — a monitor-side reliability fix that belongs in the lean Production tier.

## Scoring

- **Expected Value:** 7/10 (unlocks safe scaling + independent iteration; realized mostly at go-live)
- **Statistical Confidence:** n/a (architecture decision, not a measured finding)
- **Implementation Effort:** Medium (package carve) → High (full two-tier deploy)
- **Architectural Fit:** 10/10 — formalizes the dump-based research isolation already in use

**Comment by magedzamzam** [comment]
**Architecture review 2026-07-13 — concrete event-loop evidence for this ADR + a cheap interim mitigation.**

`spawn_bg` (`tasks.py:16`) is `asyncio.create_task` — **same event loop, single thread**. So the analytics/structure compute is off the *latency-to-placement* path (it's spawned after order placement / on a weekly cadence) but it still runs **synchronously on the executor and monitor event-loop thread**, momentarily blocking that loop's other coroutines (durable-queue consumer, `_redrive_stranded_signals`, heartbeats) for the duration of the compute:

- `analysis/estimators.py` — `hurst_rs` R/S loop (`:33`), `kalman_slope` (`:113`), and the kNN distance ranking over **all** historical candidates (`:241-249`, O(n·dims), grows unbounded with history) run sync; `sidecar.run_estimators` (`:108`) only `await`s if the estimator is awaitable, so these never yield.
- `analysis/structure_map.py:75,80,131` — `analyze_timeframe` (zigzag/label/fib) + `cluster_levels` run sync per timeframe×symbol on the monitor loop.
- `ta/capture.py:83` — `compute_timeframe` (ADX/MACD/Bollinger/…) sync between `get_bars` awaits.

**Today this is negligible** (bounded pure-Python over ~200 bars). It becomes material exactly when this ADR's trigger fires — the Phase-2/3 sidecar (#54 HMM, #55 XGBoost/LightGBM, #56 microstructure) adds heavy numpy/ML, and kNN's cost climbs with history.

**Interim mitigation available _before_ the full package split:** wrap heavy estimators in `asyncio.to_thread` (the codebase already does this in `notifications/senders.py:91`). That keeps the executor/monitor loops responsive under the ML phases without waiting for the physical Production/Research tier separation. Suggest treating "estimator compute must not run inline on the trading event loop" as an explicit precondition/checklist item for the sidecar ML phases under this ADR.

_— scheduled architecture review; full report at `beacon-claude-ai/trading-bot-tech/reports/arch-review_20260713.md`_
